firstuseauthenticator/firstuseauthenticator.py

```diff
@@ -6,15 +6,63 @@
 locally in a dbm file, and checked next time they log in.
 """
 import dbm
+import os
+from jinja2 import ChoiceLoader, FileSystemLoader
 from jupyterhub.auth import Authenticator
+from jupyterhub.handlers import BaseHandler
 from jupyterhub.orm import User
 
-from tornado import gen
+from tornado import gen, web
 from traitlets.traitlets import Unicode, Bool
 
 import bcrypt
 
 
+TEMPLATE_DIR = os.path.join(os.path.dirname(__file__), 'templates')
+
+
+class ResetPasswordHandler(BaseHandler):
+    """Render the reset password page."""
+    def __init__(self, *args, **kwargs):
+        self._loaded = False
+        super().__init__(*args, **kwargs)
+
+    def _register_template_path(self):
+        if self._loaded:
+            return
+
+        self.log.debug('Adding %s to template path', TEMPLATE_DIR)
+        loader = FileSystemLoader([TEMPLATE_DIR])
+
+        env = self.settings['jinja2_env']
+        previous_loader = env.loader
+        env.loader = ChoiceLoader([previous_loader, loader])
+
+        self._loaded = True
+
+    @web.authenticated
+    async def get(self):
+        self._register_template_path()
+        html = self.render_template('reset.html')
+        self.finish(html)
+
+    @web.authenticated
+    async def post(self):
+        data = {}
+        for arg in self.request.arguments:
```

Comment:
We only need the user's new password here. It should also be a body parameter only (comes in the body of the request), rather than a query parameter (comes in URL, like &password=new). This is important to reduce the attack surface for user's passwords being reset without their knowledge.

So, I'd recommend:

1. Using [get_body_argument](https://www.tornadoweb.org/en/stable/web.html#tornado.web.RequestHandler.get_body_argument) instead of get_argument
2. Explicitly only checking for the `password` field
3. Change the reset_password method in FirstUseAuthenticator to take a user object & password string, rather than a dictionary 'data'.

Reply:
I did this, and indeed it got much better.. much cleaner! Thank you for your feedback